File: 1636-sort-array-by-increasing-frequency/1636-sort-array-by-increasing-frequency.java
class Solution {
    public int[] frequencySort(int[] nums) {
             // Step 1: Count the frequency of each element
        Map<Integer, Integer> frequencyMap = new HashMap<>();
        for (int num : nums) {
            frequencyMap.put(num, frequencyMap.getOrDefault(num, 0) + 1);
        }

        // Step 2: Convert the array to a list of integers
        List<Integer> list = new ArrayList<>();
        for (int num : nums) {
            list.add(num);
        }

        // Step 3: Sort the list using a custom comparator
        Collections.sort(list, (a, b) -> {
            int freqA = frequencyMap.get(a);
            int freqB = frequencyMap.get(b);
            if (freqA != freqB) {
                return freqA - freqB; // Sort by frequency (ascending)
            } else {
                return b - a; // Sort by value (descending)
            }
        });

        // Step 4: Convert the list back to an array
        for (int i = 0; i < list.size(); i++) {
            nums[i] = list.get(i);
        }

        return nums;
        
    }
}

File: 1636-sort-array-by-increasing-frequency/1636-sort-array-by-increasing-frequency.py
class Solution:
    def frequencySort(self, nums: List[int]) -> List[int]:
        frequency = {}
        for num in nums:
            if num in frequency:
                frequency[num] += 1
            else:
                frequency[num] = 1
    
    # Step 2: Sort the array using the custom sorting function
        sorted_nums = sorted(nums, key=lambda x: (frequency[x], -x))
    
        return sorted_nums
        